message_ix/tests/test_soft_constraints.py

```diff
@@ -0,0 +1,61 @@
+import numpy as np
+import pandas as pd
+import pandas.testing as pdt
+
+from message_ix.testing import make_westeros
+
+
+def test_soft_constraint(test_mp):
+    s = make_westeros(test_mp, quiet=True)
+    s.check_out()
+
+    # Reduce dynamic constraint to require
+    # soft_constraint to make a feasible scenario.
+    # Chnage from 10% t0 5%.
+    df = s.par("growth_activity_up")
+    df.value = 0.05
+    s.add_par("growth_activity_up", df)
+    s.init_var("ACT_UP", ["node", "technology", "year", "time"])
+
+    # Add soft constraints for 'coal_ppl'
+    df = pd.DataFrame(
+        {
+            "node_loc": "Westeros",
+            "technology": "coal_ppl",
+            "year_act": 700,
+            "time": "year",
+            "value": [0.7],
+            "unit": "-",
+        }
+    )
+    s.add_par("soft_activity_up", df)
+
+    s.commit("Soft constraints added")
```

Comment:
If I understand this correctly, the soft-constraints do not have any (additional) costs?

Reply:
Costs have been added